[mlir][spirv] Fix int type declaration duplication when serializing

davidegrohmann opened the pull request:
At the MLIR level unsigned integer and signless integers are different types.  Indeed when looking up the two types in type definition cache they do not match.

Hence when translating a SPIR-V module which contains both usign and signless integers will contain the same type declaration twice (something like OpTypeInt 32 0) which is not permitted in SPIR-V and such generated modules fail validation.

This patch solves the problem by mapping unisgned integer types to singless integer types before looking up in the type definition cache.

davidegrohmann committed:
[mlir][spirv] Fix int type declaration duplication when serializing

At the MLIR level unsigned integer and signless integers are different
types.  Indeed when looking up the two types in type definition cache
they do not match.

Hence when translating a SPIR-V module which contains both usign and
signless integers will contain the same type declaration twice
(something like OpTypeInt 32 0) which is not permitted in SPIR-V and
such generated modules fail validation.

This patch solves the problem by mapping unisgned integer types to
singless integer types before looking up in the type definition cache.

Signed-off-by: Davide Grohmann <davide.grohmann@arm.com>
Change-Id: I9b5ecc72049d2642308b70acc335b6dc6bf6b5be

kuhar reviewed (commented):
Would it be possible to add a test for this?

davidegrohmann committed:
Resolve review comments

Also add capabilities and entry point to allow the spirv assembly
generated from the constants.mlir test to pass validation

Signed-off-by: Davide Grohmann <davide.grohmann@arm.com>
Change-Id: I22df9051de66a3dcb11d70bb95d06801aea62942

davidegrohmann commented on mlir/lib/Target/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -446,6 +446,19 @@ LogicalResult Serializer::processType(Location loc, Type type,
 Serializer::processTypeImpl(Location loc, Type type, uint32_t &typeID,
                             SetVector<StringRef> &serializationCtx) {
+
+  // Map unisgned integer types to singless integer types
```
Resolved

kuhar commented:
> The only way to test this properly is to add a validation step on the produced SPIRV assembly, but that cannot be achieve easily at the moment.

I thought we have some tests that use spir-v when present: https://github.com/llvm/llvm-project/blob/e4060d3beab3b525b49baaa15484e3c595740a2f/llvm/test/CodeGen/SPIRV/empty.ll#L2

Could we do something similar in mlir spirv serialization?

If this is not straightforward, validating by hand is also fine IMO.

davidegrohmann committed:
Add optional SPIR-V validation in constant.mlir Target test

SPIR-V validation is run if spirv-tools are available otherwise
skipped.

Signed-off-by: Davide Grohmann <davide.grohmann@arm.com>
Change-Id: Id0af7c983c4bd6596a4837812f22ced3c7616177

davidegrohmann commented:
> I thought we have some tests that use spir-v when present:
> 
> https://github.com/llvm/llvm-project/blob/e4060d3beab3b525b49baaa15484e3c595740a2f/llvm/test/CodeGen/SPIRV/empty.ll#L2
> 
> Could we do something similar in mlir spirv serialization?
> 
> If this is not straightforward, validating by hand is also fine IMO.

Thanks for the suggestion I have added that to the test. I have also validated manually the SPIR-V module produced by that test and it passes locally for me.

kuhar commented:
> Thanks for the suggestion I have added that to the test. I have also validated manually the SPIR-V module produced by that test and it passes locally for me.

Have you checked that the spirv tools invocation works as expected and fails to validate without your fix? I haven't tried it myself, so I'm not sure if it has to be enabled somewhere in cmake etc.

davidegrohmann commented:
> Have you checked that the spirv tools invocation works as expected and fails to validate without your fix? I haven't tried it myself, so I'm not sure if it has to be enabled somewhere in cmake etc.

I was hoping that to work out of the box since there are also other tests in the same folder using the same spirv-tools configuration, for example: `mlir/test/Target/SPIRV/consecutive-selection.spv` and `mlir/test/Target/SPIRV/selection.spv`

In any case I have tried to make it work locally without success. I have built everything enabling the SPIRV target and enable the SPIRV tests by adding `-DLLVM_TARGETS_TO_BUILD='host;SPIRV'` and `-DLLVM_INCLUDE_SPIRV_TOOLS_TESTS=ON`

Full command:
```
cmake -G Ninja -B build -S llvm
    -DLLVM_ENABLE_PROJECTS=mlir
    -DLLVM_TARGETS_TO_BUILD='host;SPIRV'
    -DCMAKE_BUILD_TYPE=Release
    -DLLVM_REQUIRES_RTTI=ON
    -DLLVM_REQUIRES_EH=ON
    -DLLVM_ENABLE_RTTI=ON
    -DLLVM_ENABLE_EH=ON
    -DLLVM_ENABLE_ASSERTIONS=ON
    -DLLVM_ABI_BREAKING_CHECKS=FORCE_OFF
    -DMLIR_INCLUDE_INTEGRATION_TESTS=ON
    -DLLVM_ENABLE_TERMINFO=OFF
    -DLLVM_INCLUDE_SPIRV_TOOLS_TESTS=ON
```

Any ideas what am I missing?

In any case I have manually tested that the command `mlir-translate -no-implicit-module -serialize-spirv %s | spirv-val` works for `mlir/test/Target/SPIRV/constant.mlir`

davidegrohmann committed:
Merge branch 'main' into fix-unsign-signless-type-serialization-duplication

kuhar commented:
> Any ideas about what might I be missing?

I'd try to use the git history and see how this was initially plumbed through in clang tests. Once we figure this out, it will be very valuable in other tests that deal with (de)serialization. If some cmake configuration is required, we can document it under https://mlir.llvm.org/getting_started/TestingGuide/ .

> -DLLVM_TARGETS_TO_BUILD='host;SPIRV'

This looks unrelated to me.

davidegrohmann committed:
Enable spirv-tools integration when running mlir Target tests

Also fix validation failures caused by a rebase

Signed-off-by: Davide Grohmann <davide.grohmann@arm.com>
Change-Id: I221881875c77a30a4df4bd9b15288ca10e9c8e9c

davidegrohmann commented:
@kuhar is there anything more that is required for this PR to be merged?

kuhar commented:
> @kuhar is there anything more that is required for this PR to be merged?

Merged. @davidegrohmann could you submit a follow up PR and update the mlir documentation to explain how to run tests that use spirv tools and also decouple the cmake flag for spirv tools from the flag that enables the spirv backend?

joker-eph commented on mlir/test/lit.local.cfg:
```diff
@@ -0,0 +1,7 @@
+if not "SPIRV" in config.root.targets:
+    config.unsupported = True
```
Why do we need the SPIRV backend to run MLIR unit-tests?

davidegrohmann replied:
Please see this comment: https://github.com/llvm/llvm-project/pull/143108#issuecomment-2963067688

kuhar replied:
I missed this -- if this disables all mlir tests, we should revert the PR

kuhar replied:
Revert PR: https://github.com/llvm/llvm-project/pull/144773
